add smoke tests + fix service worker tests

samccone opened the pull request:
https://travis-ci.org/GoogleChrome/lighthouse/jobs/122697303#L1114
^ Interesting... looks like we are not picking up the service worker at all ... curious if anyone has insight here... could it be due to the revert of the force update I did? Is something else going on? 

---

Looks like the force update of service worker API changed a bit today
https://travis-ci.org/GoogleChrome/lighthouse/jobs/122689130#L2606

Which is causing the test to fail until I reverted the commit that enabled this behavior

brendankenny commented:
when you wake up, this is now failing :)

brendankenny commented:
This is a great start, and I agree this starts to test a ton of integration stuff (not to mention driver/gatherers) we don't test at all today.

brendankenny commented on run-smoke-tests.sh:
```diff
@@ -0,0 +1,19 @@
+
+./cli.js http://localhost:9999/online-only.html > results
+
+if ! grep -q "URL responds with a 200 when offline: false" results; then
```
would much rather test non-pretty-printed results than strings, as the first is more likely to be stable over time than the second, and easier for humans to quickly scan and keep the test expectations up to date. We can do that in a follow up PR if you want, though.

samccone replied:
:+1: started doing a version using jq . to parse the response but since it is just an array with objects and things are not keyed inside of an object, the matching logic got much worse and uglier than this.

:+1: to iterating more

brendankenny commented on gather-scheduler.js:
```diff
@@ -51,9 +51,6 @@ class GatherScheduler {
       // Clean all browser caches.
       .then(_ => driver.cleanAndDisableBrowserCaches())
 
-      // Force SWs to update on load.
```
@paulirish or @paullewis for what's going on here with the new errors

paullewis replied:
Checking this out now.

paullewis commented:
In `run-smoke-tests.sh`:
```diff
+ sleep 5s
```

paullewis commented:
And switch to ToT with:

```diff
- 'https://download-chromium.appspot.com/dl/Linux_x64?type=continuous'
+ 'https://download-chromium.appspot.com/dl/Linux_x64?type=snapshots'
```

samccone commented:
so far I have moved to launching chrome directly into the file and removed it from a background process, so nothing is still working. I then went and moved the order of the tests around thinking that it could be due to rerunning the tests in the same chrome instance.. still no good :goat: 

My next steps will be to debug this on my linux desktop and see if i can reproduce the issue. My current thinking is that there may be something weird going on with focus and chrome that is causing the service worker events not to fire... like perhaps since it is on a different (virtual) screen :monkey: it is not even thinking it needs to do any work... but these are just some ramblings... I will need to see if I can repo...

Any other insights that anyone has would be very much appreciated :clock11: :skull:

samccone commented:
![image](https://cloud.githubusercontent.com/assets/883126/14550803/ba1b268a-0280-11e6-808f-4bbaa02693d6.png)


via a clean profile ^

note we get started but no actived

samccone commented:
Just tested with the extension manually, seems like everything is still in working order.

👀 welcome on this one.

**Just a note**, ⚠️ squashing this PR would not be advisable ⚠️

brendankenny commented on test/download-chrome.sh:
```diff
@@ -12,5 +12,5 @@ if [ -e "$LIGHTHOUSE_CHROMIUM_PATH" ]
   echo "cached chrome found"
 else
-  wget 'https://download-chromium.appspot.com/dl/Linux_x64?type=continuous' -O chrome.zip && unzip chrome.zip
+  wget 'https://download-chromium.appspot.com/dl/Linux_x64?type=snapshots' -O chrome.zip && unzip chrome.zip
```
wasn't the problem not continuous vs snapshots in the end? Can we keep using the continuous build?

samccone replied:
we still need the latest for the new features we are using in 52

brendankenny replied:
👍

brendankenny commented on helpers/browser/driver.js:
```diff
@@ -150,35 +150,33 @@ class ChromeProtocol {
-      .then(response => {
+      return this.sendCommand('Page.navigate', {url});
+    }).then(response => {
+      return new Promise((res, _) => {
```
`(resolve, reject)`

samccone replied:
done

brendankenny commented on helpers/browser/driver.js:
```diff
@@ -150,35 +150,33 @@ class ChromeProtocol {
-          return this.sendCommand('Page.reload', {ignoreCache: true});
-        }
+    return this.sendCommand('Page.enable').then(_ => {
+      return this.sendCommand('Page.getNavigationHistory');
```
prefer the original
```
.then(_ => this.sendCommand('Page.getNavigationHistory'))
```
formatting for this line

samccone replied:
done

brendankenny commented on helpers/browser/driver.js:
```diff
@@ -150,35 +150,33 @@ class ChromeProtocol {
+      return new Promise((res, _) => {
         this.url = url;
 
         if (!waitForLoaded) {
```
why do we even have `waitForLoaded`?

samccone replied:
duno, let's open an issue and take a look

brendankenny commented on gatherers/offline.js:
```diff
@@ -102,22 +102,21 @@ class Offline extends Gather {
-        .then(_ => {
-          this.artifact = {responseCode};
-        });
+      return Offline.getOfflinePageStatus(driver).then(ret => {
```
this is fine, but does mess up git blame in exchange for arguable readability

samccone replied:
ack

brendankenny commented on gather-scheduler.js:
```diff
@@ -18,99 +18,114 @@
+    const args = Array.from(arguments).slice(1);
+
+    return fns.reduce((prev, curr) => {
+      return prev.then(_ => curr.apply(null, args));
```
`curr(...args)`

samccone replied:
nice 👍

brendankenny commented on gather-scheduler.js:
```diff
@@ -18,99 +18,114 @@
+  static run(gatherers, options) {
+    const driver = options.driver;
+    const tracingData = {};
+    const self = GatherScheduler;
```
the combination of using `self` to call `static` methods is definitely a code smell. While you're in here, just drop the class and have the static methods be functions

samccone replied:
it is just a shorthand for the class, i will rename to klass

brendankenny replied:
that has the same issue

samccone replied:
ok let's do this in another PR seems like a good abstraction / improvement.. or at least one that we can play around with in another PR sandbox

brendankenny commented on gather-scheduler.js:
```diff
@@ -18,99 +18,114 @@
-      // Gather: tearDown phase.
-      .then(_ => this._runPhase(gatherers,
-        gatherer => gatherer.tearDown(options, tracingData)))
+    if (options.url === undefined || options.url === null) {
```
why would the url be `null`?

samccone replied:
unknown, just handling the case :)

I did not wanted to be explicit here because if someone passed `''` I would not want to fail

brendankenny commented on gather-scheduler.js:
```diff
@@ -18,99 +18,114 @@
+  }
 
+  static setupDriver(driver, gatherers, options) {
+    return new Promise((res, _) => {
```
`(resolve, reject)`

brendankenny replied:
or just the original version, returning `driver.beginEmulation()` or `Promise.resolve()` directly

samccone replied:
since there is some logic inside of the block here I wanted to wrap it all in a promise constructor to handle the case where something could potentially throw 💥

brendankenny commented on gather-scheduler.js:
```diff
@@ -18,99 +18,114 @@
+    return GatherScheduler.runSeries([
+      driver.connect.bind(driver),
+      self.setupDriver,
+      self._phase('setup'),
```
I'm still not wild about this :) Calling the phase methods with a string accessor seems like a symptom of impedance mismatch in two areas:
- `GatherScheduler` is really a collection of static functions but is calling into real gatherer classes, which means you can't just call `runSeries` over the different phases; you need something for the `this` binding
- Pretending these functions are easily composable because they're all endomorphisms over the same type when they actually aren't. Taking the union of domains is one way to force them in that round hole, but it can easily end with an unstructured mess of arguments being passed in (this would be even more obvious if `options` and `tracingData` were split into their individual properties before being passed into `runSeries`). And because our functions aren't side-effect free, you lose information at the call site of what a function is changing, relying on having to inspect the function itself.

In both cases you trade some clarity and static verifiability for an easier to read list of steps. For purity's sake we'd want to make them actually endomorphisms/monadic, instead of having the state repeatedly reinjected, or if it's the readability we're after we actually want `async`/`await` here. The first approach is probably overkill, the second isn't possible yet.

I personally feel that something closer to the existing `then`-heavy structure, with the driver steps pulled out as you have here, is a better expression of what we're doing here while we're waiting on `await` support to drop, rather than making a pipeline over (conceptually-)orthogonal domains.

brendankenny replied:
As an example of why the second bullet point isn't ideal: I've been thinking about nice ways we can get away from the magic `this.artifact` on gatherers, returning a promise that resolves to the artifact at some stage. This would require adding an `artifacts` array to the arguments passed into `runSeries`, which means future readers would have to go through all the called functions to find out that only calls to `runPhase` was creating possible artifacts.

brendankenny replied:
This is, incidentally, why I was hoping for separate PRs for the addition of reload phases vs the scheduler refactor :)

samccone replied:
we can iterate on this more, I felt that this was an OK first step.. you are correct none of these are side effect free or re-entrant (however I do not think these can ever be since we do not have state free exits). The goal here was the make the list of steps readable, and I feel pretty ok with what this ended up being like.

I felt that adding the new steps onto the .then chain pushed it over the limit of readability.


I want to iterate more on this and we can continue to tweak

samccone replied:
you lose information at the call site of what a function is changing, relying on having to inspect the function itself

^ I think it was a lot like this before, it was all just a really long .then chain and hard to grok what was going on and what the flow was.

samccone replied:
> which means future readers would have to go through all the called functions to find out that only calls to runPhase was creating possible artifacts.

yep, it is an abstraction, I personally am willing to pay this cost for the current layout of the flow code.

samccone replied:
> This is, incidentally, why I was hoping for separate PRs for the addition of reload phases vs the scheduler refactor :)

I did individual commits to make moving / reverting easy

as I said above I found layering on more .then clauses unacceptable

samccone commented:
Let me know what you want me to do, I can revert all of this if you think this is a no go.

paullewis commented:
The main issues here are around the refactor of the gatherer. Can we land this with the current gatherer logic (with the addition of the second page load) and deal with a gatherer refactor in a separate PR? I would like the SW tests to be fixed sooner rather than later.

paullewis commented on gather-scheduler.js:
```diff
@@ -18,99 +18,124 @@
+    }
 
+    return driver.connect().then(_ => {
+      return GatherScheduler.setupDriver(driver, gatherers, options);
```
Why the explicit returns?

samccone replied:
multi-line =>

paullewis replied:
Yeah, but whyyyyyyy?

samccone replied:
lines were too long

paullewis replied:
👍

brendankenny replied:
that's why I was saying make the methods just functions (since they're already operating as ones). Then you can just do, like,
```
    return driver.connect()
      .then(_ => setupDriver(driver, gatherers, options))
      .then(_ => runPhase(gatherers, gatherer => gatherer.setup(options)))
      .then(_ => beginPassiveCollection(driver))
      .then(_ => runPhase(gatherers, gatherer => gatherer.beforePageLoad(options)))
      .then(_ => loadPage(driver, gatherers, options))
      .then(_ => runPhase(gatherers, gatherer => gatherer.afterPageLoad(options)))
      .then(_ => endPassiveCollection(driver, tracingData))
      .then(_ => runPhase(gatherers,
          gatherer => gatherer.afterTraceCollected(options, tracingData)))
      .then(_ => runPhase(gatherers, gatherer => gatherer.reloadSetup(options)))
      .then(_ => runPhase(gatherers, gatherer => gatherer.beforeReloadPageLoad(options)))
      .then(_ => reloadPage(driver, options))
      .then(_ => runPhase(gatherers, gatherer => gatherer.afterReloadPageLoad(options)))
      .then(_ => driver.disconnect())
      .then(_ => runPhase(gatherers, gatherer => gatherer.tearDown(options)))
      .then(_ => {
        // Collate all the gatherer results.
        return gatherers.map(g => g.artifact).concat(
            {networkRecords: tracingData.networkRecords},
            {traceContents: tracingData.traceContents});
      });
```

brendankenny replied:
just to be clear, this can be held off for a future PR, though :)

samccone replied:
we can iterate on it more

paulirish replied:
Line length. :3

brendankenny commented on gather-scheduler.js:
```diff
@@ -18,99 +18,124 @@
-      Promise.resolve()
-    );
+  static runSeries(fns) {
+    const args = Array.from(arguments).slice(1);
```
it'll be so nice when we get rest parameters in node as well

brendankenny replied:
I think you want to remove this, actually, since you aren't using it in the current changes

samccone replied:
good catch, fixed

brendankenny commented on audits/mobile-friendly/display.js:
```diff
@@ -55,10 +55,13 @@ class ManifestDisplay extends Audit {
-      ManifestDisplay.hasRecommendedValue(manifest.display.value),
-      manifest.display.value,
+      hasRecommendedValue,
+      displayValue,
```
the manifest parser also has a debug string for display value if it isn't one of `fullscreen`, `standalone`, `minimal-ui`, or `browser` it might be worth piping through. Not sure the best way to combine it with this warning, though.

samccone replied:
open up an issue to track we can improve it there

brendankenny commented:
Nice! 💣 => 💥 

I'm sorry I make all things painful 🐗
